File: firmware/src/core/protocol.cpp
#include "protocol.h"

uint8_t proto_checksum(uint8_t channel, uint8_t device_id, uint8_t len,
                       const uint8_t *payload)
{
    uint8_t chk = 0;
    chk ^= channel;
    chk ^= device_id;
    chk ^= len;
    for (uint8_t i = 0; i < len; i++) {
        chk ^= payload[i];
    }
    return chk;
}

size_t proto_build(uint8_t channel, uint8_t device_id,
                   const uint8_t *payload, uint8_t len,
                   uint8_t *out, size_t out_cap)
{
    // Frame overhead: SOF, CH, DEV, LEN, CHK, EOF = 6 bytes.
    const size_t total = (size_t)len + 6;
    if (out == nullptr || out_cap < total) return 0;
    if (len > 0 && payload == nullptr) return 0;

    size_t i = 0;
    out[i++] = PROTO_SOF;
    out[i++] = channel;
    out[i++] = device_id;
    out[i++] = len;
    for (uint8_t p = 0; p < len; p++) out[i++] = payload[p];
    out[i++] = proto_checksum(channel, device_id, len, payload);
    out[i++] = PROTO_EOF;
    return i;
}

ProtocolParser::ProtocolParser()
{
    _cb = nullptr;
    _good = 0;
    _bad = 0;
    reset();
}

void ProtocolParser::reset()
{
    _state = S_WAIT_SOF;
    _payload_idx = 0;
    _chk_calc = 0;
    _chk_recv = 0;
    _pkt.channel = 0;
    _pkt.device_id = 0;
    _pkt.len = 0;
}

void ProtocolParser::feed(const uint8_t *data, size_t n)
{
    if (!data) return;
    for (size_t i = 0; i < n; i++) feed(data[i]);
}

bool ProtocolParser::feed(uint8_t b)
{
    switch (_state) {
    case S_WAIT_SOF:
        if (b == PROTO_SOF) {
            _state = S_CHANNEL;
            _chk_calc = 0;
        }
        break;

    case S_CHANNEL:
        _pkt.channel = b;
        _chk_calc ^= b;
        _state = S_DEVICE;
        break;

    case S_DEVICE:
        _pkt.device_id = b;
        _chk_calc ^= b;
        _state = S_LEN;
        break;

    case S_LEN:
        _pkt.len = b;
        _chk_calc ^= b;
        _payload_idx = 0;
        // LEN is a single byte so it can never exceed PROTO_MAX_PAYLOAD,
        // but guard anyway for buffer safety.
        if (_pkt.len > PROTO_MAX_PAYLOAD) {
            _bad++;
            reset();
            break;
        }
        _state = (_pkt.len == 0) ? S_CHK : S_PAYLOAD;
        break;

    case S_PAYLOAD:
        if (_payload_idx < PROTO_MAX_PAYLOAD) {
            _pkt.payload[_payload_idx] = b;
        }
        _chk_calc ^= b;
        _payload_idx++;
        if (_payload_idx >= _pkt.len) {
            _state = S_CHK;
        }
        break;

    case S_CHK:
        _chk_recv = b;
        _state = S_EOF;
        break;

    case S_EOF:
        if (b == PROTO_EOF && _chk_recv == _chk_calc) {
            _good++;
            if (_cb) _cb(&_pkt);
            reset();
            return true;
        } else {
            _bad++;
            reset();
            // The byte that broke the frame might itself be a new SOF.
            if (b == PROTO_SOF) {
                _state = S_CHANNEL;
                _chk_calc = 0;
            }
        }
        break;

    default:
        reset();
        break;
    }
    return false;
}

File: pi/protocol.py
"""Binary protocol framing — must match firmware/src/core/protocol.*

Frame: [0xAA][CHANNEL][DEVICE_ID][LEN][PAYLOAD...][CHK][0x55]
Checksum: XOR of CHANNEL + DEVICE_ID + LEN + PAYLOAD bytes.
"""

SOF = 0xAA
EOF = 0x55

MAX_PAYLOAD = 255

# Channels
CH_CONTROL = 0x01
CH_STATUS = 0x02
CH_STREAM = 0x03
CH_OTA = 0x04
CH_ROUTING = 0x05


class Packet:
    __slots__ = ("channel", "device_id", "payload")

    def __init__(self, channel: int, device_id: int, payload: bytes):
        self.channel = channel & 0xFF
        self.device_id = device_id & 0xFF
        self.payload = bytes(payload)

    def __repr__(self):
        return f"Packet(ch=0x{self.channel:02X}, dev=0x{self.device_id:02X}, len={len(self.payload)})"


def checksum(channel: int, device_id: int, length: int, payload: bytes) -> int:
    chk = (channel ^ device_id ^ length) & 0xFF
    for b in payload:
        chk ^= b
    return chk & 0xFF


def build_frame(channel: int, device_id: int, payload: bytes) -> bytes:
    payload = bytes(payload)
    length = len(payload)
    if length > MAX_PAYLOAD:
        raise ValueError("payload exceeds 255 bytes")
    chk = checksum(channel, device_id, length, payload)
    return bytes([SOF, channel & 0xFF, device_id & 0xFF, length]) + payload + bytes([chk, EOF])


class Parser:
    """Streaming, byte-at-a-time state machine matching the firmware parser."""

    S_SOF, S_CH, S_DEV, S_LEN, S_PAYLOAD, S_CHK, S_EOF = range(7)

    def __init__(self):
        self.reset()
        self.good = 0
        self.bad = 0

    def reset(self):
        self._state = self.S_SOF
        self._channel = 0
        self._device = 0
        self._len = 0
        self._payload = bytearray()
        self._chk_calc = 0
        self._chk_recv = 0

    def feed(self, data: bytes):
        """Feed bytes; yields complete validated Packet objects."""
        for b in data:
            pkt = self._feed_byte(b)
            if pkt is not None:
                yield pkt

    def _feed_byte(self, b: int):
        s = self._state
        if s == self.S_SOF:
            if b == SOF:
                self._state = self.S_CH
                self._chk_calc = 0
        elif s == self.S_CH:
            self._channel = b
            self._chk_calc ^= b
            self._state = self.S_DEV
        elif s == self.S_DEV:
            self._device = b
            self._chk_calc ^= b
            self._state = self.S_LEN
        elif s == self.S_LEN:
            self._len = b
            self._chk_calc ^= b
            self._payload = bytearray()
            self._state = self.S_CHK if b == 0 else self.S_PAYLOAD
        elif s == self.S_PAYLOAD:
            self._payload.append(b)
            self._chk_calc ^= b
            if len(self._payload) >= self._len:
                self._state = self.S_CHK
        elif s == self.S_CHK:
            self._chk_recv = b
            self._state = self.S_EOF
        elif s == self.S_EOF:
            if b == EOF and self._chk_recv == self._chk_calc:
                self.good += 1
                pkt = Packet(self._channel, self._device, bytes(self._payload))
                self.reset()
                return pkt
            else:
                self.bad += 1
                self.reset()
                if b == SOF:
                    self._state = self.S_CH
                    self._chk_calc = 0
        return None
